cratetorrent/src/peer/codec.rs

```diff
@@ -417,19 +406,16 @@ impl Decoder for PeerCodec {
             return Ok(None);
         }
 
-        // hack:
-        // `get_*` integer extractors consume the message bytes by advancing
-        // buf's internal cursor. However, we don't want to do this as at this
-        // point we aren't sure we have the full message in the buffer, and thus
-        // we just want to peek at this value.
-        //
-        // However, this is not supported by the `bytes` crate API
-        // (https://github.com/tokio-rs/bytes/issues/382), so we need to work
-        // this around by getting a reference to the underlying buffer and
-        // performing the message length extraction on the returned slice, which
-        // won't advance `buf`'s cursor
-        let mut tmp_buf = buf.bytes();
-        let msg_len = tmp_buf.get_u32() as usize;
+        let msg_len = if let Some(len_bytes) = buf.get(0..4) {
```

Comment:
Perhaps using `std::io::Cursor` as described in https://tokio.rs/tokio/tutorial/framing#parsing to keep the `get_*` API would be more straightforward for these peek ahead checks.

```
let mut tmp_buf = Cursor::new(&buf);
let msg_len = tmp_buf.get_u32() as usize;
```

Reply:
Agreed, this looks more straightforward. Also, keeping a brief comment on why the temporary buffer would be useful for future maintainers.